src/smartWalletDiscovery.ts

```diff
@@ -0,0 +1,321 @@
+
+// TODO: copied from @ethersproject/wordlists/src.ts/wordlists.ts as that doesn't allow for locale type extraction. Remove (including imports) when this is fixed (https://github.com/ethers-io/ethers.js/pull/3514).
+const wordlists = {
+  cz,
+  en,
+  es,
+  fr,
+  it,
+  ja,
+  ko,
+  zh: zh_cn,
+  zh_cn,
+  zh_tw,
+} as const;
+
+// TODO: check that the optional params have to be optional
+type Config = {
+  eoaGap: number;
+  sWalletGap: number;
+  searchBalance: boolean;
+  searchNonce: boolean;
+  searchTokenBalance: boolean;
+  searchDeployEvents: boolean;
+  network: string; //TODO: potentially restictable by program-wide config
+  factory: string;
+  isCustomWallet: boolean;
+  mnemonicLanguage: Locale;
+  recoverer: string;
+  logic: string;
+  logicParamsHash: string;
+  reconnectionAttempts: number;
+  tokens: string[];
+};
+
+const DEFAULT_DISCOVERY_CONFIG: Partial<Config> = {
+  eoaGap: 20,
+  sWalletGap: 20,
+  searchBalance: true,
+  searchNonce: true,
+  searchTokenBalance: true,
+  searchDeployEvents: true,
+  network: 'mainnet', //FIXME: this should probably be something else. Also should be linked/mapped to the network id or address defined in program-wide config or env vars;
+  mnemonicLanguage: 'en',
+  recoverer: constants.AddressZero,
+  tokens: [],
+};
+
+//Account discovery https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki#account-discovery
+const DEFAULT_DERIVATION_PATH_PURPOSE = 44;
+const RSK_COIN_TYPE = 137; // src: https://github.com/satoshilabs/slips/blob/master/slip-0044.md
+const RSK_TESTNET_COIN_TYPE = 37310; // src: https://github.com/rsksmart/RSKIPs/blob/master/IPs/RSKIP57.md
+const DEFAULT_DERIVATION_PATH_CHANGE = 0;
+
+const constructDerivationPath = (
+  addressIdx: number,
+  hardenedIdx: number,
+  coinType?: number,
+  isTestNet = false,
+  purpose = DEFAULT_DERIVATION_PATH_PURPOSE,
+  change: 0 | 1 = DEFAULT_DERIVATION_PATH_CHANGE
+) => {
+  return `m/${purpose}'/${
+    coinType || isTestNet ? RSK_TESTNET_COIN_TYPE : RSK_COIN_TYPE
+  }'/${hardenedIdx}'/${change}/${addressIdx}`;
+};
+
+const getSWAddress = (
+  config: Config,
+  ownerEOA: string,
+  walletIndex: number,
+  bytecodeHash: string
+) => {
+  const isCustom = config.logic && config.logicParamsHash;
+  const salt = isCustom
+    ? solidityKeccak256(
+        ['address', 'address', 'address', 'bytes32', 'uint256'],
+        [
+          ownerEOA,
+          config.recoverer,
+          config.logic,
+          config.logicParamsHash ?? SHA3_NULL_S,
+          walletIndex,
+        ]
+      )
+    : solidityKeccak256(
+        ['address', 'address', 'uint256'],
+        [ownerEOA, config.recoverer, walletIndex]
+      );
+
+  const data: string = solidityKeccak256(
+    ['bytes1', 'address', 'bytes32', 'bytes32'],
+    ['0xff', config.factory, salt, bytecodeHash]
+  );
+
+  const address = `0x${data.slice(26, data.length)}`;
+
+  return getAddress(address);
+};
+
+const hasAccActivity = async (
```

Comment:
I'd rename it to `hasAccountActivity` to be clearer

Reply:
I've actually renamed it to represent an action: `checkHasActivity`